extensions/iceberg/src/main/java/io/deephaven/iceberg/util/IcebergTableAdapter.java

```diff
@@ -633,46 +655,312 @@ private static TableDefinition fromSchema(
+                            .collect(Collectors.toList()));
+        }
+        final ParquetInstructions parquetInstructions = writeInstructions.toParquetInstructions(
+                onWriteCompleted, parquetTableDefinition, table.schema().idToName());
+
+        // Write the data to parquet files
+        int count = 0;
+        for (int idx = 0; idx < dhTables.size(); idx++) {
+            final Table dhTable = dhTables.get(idx);
+            if (dhTable.numColumns() == 0) {
+                // Skip writing empty tables with no columns
+                continue;
+            }
+            final String fileName = String.format(
+                    "00000-%d-%s.parquet",
+                    count++,
+                    UUID.randomUUID());
+            final String relativePath;
+            if (partitionSpec.isPartitioned()) {
+                relativePath = String.format("%s/%s", partitionPaths.get(idx), fileName);
+            } else {
+                relativePath = fileName;
+            }
+            final String newDataLocation = table.locationProvider().newDataLocation(relativePath);
+            ParquetTools.writeTable(dhTable, newDataLocation, parquetInstructions);
+        }
+        return parquetFilesWritten;
+    }
+
+    /**
+     * Commit the changes to the Iceberg table by creating a snapshot.
+     */
+    private void commit(
+            @NotNull final IcebergUtils.SpecAndSchema newSpecAndSchema,
+            @NotNull final Iterable<DataFile> appendFiles,
+            final boolean overwrite,
+            final boolean schemaVerified) {
+        final Transaction icebergTransaction = table.newTransaction();
+        final Snapshot currentSnapshot = table.currentSnapshot();
+        // For a null current snapshot, we are creating a new table. So we can just append instead of overwriting.
+        if (overwrite && currentSnapshot != null) {
+            // Fail if the table gets changed concurrently
+            final OverwriteFiles overwriteFiles = icebergTransaction.newOverwrite()
+                    .validateFromSnapshot(currentSnapshot.snapshotId())
+                    .validateNoConflictingDeletes()
+                    .validateNoConflictingData();
+
+            // Delete all the existing data files in the table
+            try (final Stream<DataFile> dataFiles = allDataFiles(table, currentSnapshot)) {
+                dataFiles.forEach(overwriteFiles::deleteFile);
+            }
+            appendFiles.forEach(overwriteFiles::addFile);
+            overwriteFiles.commit();
+
+            // Update the spec and schema of the existing table.
+            // If we have already verified the schema, we don't need to update it.
+            if (!schemaVerified) {
+                if (!table.schema().sameSchema(newSpecAndSchema.schema())) {
+                    final UpdateSchema updateSchema = icebergTransaction.updateSchema().allowIncompatibleChanges();
```

Comment:
I wrote another comment on this. We should be clear about whether we want to allow evolutions that would break a refreshing DH table. I think no.